# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [7]:
import pandas as pd
df = pd.read_parquet('emendas_parlamentares (1).parquet')
df.head()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.0,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.0,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.0,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.0,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.0,1131133.94,Sem informação


## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [15]:
df_2022 = df[df['Ano da Emenda'] == 2022]
df_2022.head(30)
df_2022.shape[0]

6108

Foram registradas 6108 emendas no ano de 2022. 

## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [17]:
filtro = (df['Localidade de aplicação do recurso'].str.contains(' - PA| - AM| - RO| - AC| - RR| - AP| - TO')) & (df['Valor Empenhado'] > 500000)
df_norte_500k = df[filtro]

df_norte_500k.head(5)
df_norte_500k.shape[0]

2067

2067 emendas atendem as condições de localidade Região Norte e valor empenhado maior que 500.000 ao mesmo tempo.

## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [20]:
df_sp = df[df['Localidade de aplicação do recurso'].str.contains(' - SP')]
df_rj = df[df['Localidade de aplicação do recurso'].str.contains(' - RJ')]

len(df_sp)
len(df_rj)

df[df['Localidade de aplicação do recurso'].str.contains(' - SP| - RJ')]['Localidade de aplicação do recurso'].str.extract('(- ..)')[0].value_counts()

0
- SP    4730
- RJ    3805
Name: count, dtype: int64

O Estado que possui mais emendas é São Paulo, contendo 4730 emendas.

## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [22]:
print(df['Código da Emenda'].value_counts())


df['Código da Emenda'] = df['Código da Emenda'].replace('Sem informação', 'Não disponível')

print(df['Código da Emenda'].value_counts())

Código da Emenda
Não disponível    17808
201833870003          2
201833870004          2
201833870006          2
201833870007          2
                  ...  
202544840005          1
202544840006          1
202544840007          1
202544840008          1
202544840009          1
Name: count, Length: 70731, dtype: int64
Código da Emenda
Não disponível    17808
201833870003          2
201833870004          2
201833870006          2
201833870007          2
                  ...  
202544840005          1
202544840006          1
202544840007          1
202544840008          1
202544840009          1
Name: count, Length: 70731, dtype: int64


A contagem de valores únicos da coluna permanece sendo 17.808.

## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [23]:
df['Nome Função'] = df['Nome Função'].str.upper()
df['Nome Função'].unique()

<ArrowStringArray>
[            'URBANISMO',   'COMÉRCIO E SERVIÇOS',                 'SAÚDE',
              'EDUCAÇÃO',           'AGRICULTURA',    'ASSISTÊNCIA SOCIAL',
      'DESPORTO E LAZER',       'DEFESA NACIONAL',            'SANEAMENTO',
 'DIREITOS DA CIDADANIA',      'GESTÃO AMBIENTAL',               'CULTURA',
  'CIÊNCIA E TECNOLOGIA',     'SEGURANÇA PÚBLICA',   'ORGANIZAÇÃO AGRÁRIA',
          'COMUNICAÇÕES',         'ADMINISTRAÇÃO',             'HABITAÇÃO',
             'INDÚSTRIA',              'TRABALHO',              'MÚLTIPLO',
            'TRANSPORTE',    'PREVIDÊNCIA SOCIAL',    'ENCARGOS ESPECIAIS',
   'RELAÇÕES EXTERIORES',               'ENERGIA',   'ESSENCIAL À JUSTIÇA']
Length: 27, dtype: str

Utilizando o método str.upper() é possível transformar todos os valores disponíveis na coluna Nome Função para que fiquem em letra maiúscula. 

## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [25]:
print(df.duplicated().sum())
df = df.drop_duplicates()
print(df.shape[0])

0
88596


Não foram encontradas linha duplicadas no dataset. O número total de linhas é de 88596.

## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro? 

In [26]:
print(df.duplicated(subset=['Código da Emenda', 'Município']).sum())
df_subset = df.drop_duplicates(subset=['Código da Emenda', 'Município'])
print(df_subset.shape[0])

12852
75744


Existem 12852 linhas duplicadas considerando apenas o par Código da Emenda e Município. Sobraram 75744 linhas após a remoção.

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [27]:
nulos_por_coluna = df.isnull().sum()
print(nulos_por_coluna)

coluna_max_nulos = nulos_por_coluna.idxmax()
total_max_nulos = nulos_por_coluna.max()
print(coluna_max_nulos, total_max_nulos)

Código da Emenda                          0
Ano da Emenda                             0
Tipo de Emenda                            0
Código do Autor da Emenda             15961
Nome do Autor da Emenda               15961
Número da emenda                          0
Localidade de aplicação do recurso        0
Código Município IBGE                     0
Município                                 0
Código UF IBGE                            0
UF                                        0
Região                                    0
Código Função                             0
Nome Função                               0
Código Subfunção                          0
Nome Subfunção                            0
Código Programa                           0
Nome Programa                             0
Código Ação                               0
Nome Ação                                 0
Código Plano Orçamentário                 0
Nome Plano Orçamentário                   0
Valor Empenhado                 

As colunas Código do Autor da Emenda e Nome do Autor da Emenda são as que possuem maior volume de dados faltantes no arquivo, com 15.961 linhas vazias. Todas as colunas que aparecem com 0 estão totalmente preenchidas.

## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [29]:
df['Código do Autor da Emenda'] = df['Código do Autor da Emenda'].fillna(-1)
df['Nome do Autor da Emenda'] = df['Nome do Autor da Emenda'].fillna('desconhecido')

print(df[['Código do Autor da Emenda', 'Nome do Autor da Emenda']].isnull().sum())

Código do Autor da Emenda    0
Nome do Autor da Emenda      0
dtype: int64


os 15.961 espaços que antes estavam vazios agora contêm -1 (no código) ou "desconhecido" (no nome). O dataset está preenchido nessas colunas.

## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [30]:
print(df['Ano da Emenda'].dtype)
df['Ano da Emenda'] = df['Ano da Emenda'].astype(str)
print(df['Ano da Emenda'].dtype)

int64
str


Antes: Os números eram tratados como números inteiros.
Depois: Após convertido para String, o tipo aparece como object. Isso significa que o ano é tratado como um texto.

## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [31]:
df['Valor Empenhado'] = df['Valor Empenhado'].astype(float)
print(df['Valor Empenhado'].dtype)

float64


Após a conversão, o tipo da coluna Valor Empenhado é float.

## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [32]:
total_por_regiao = df.groupby('Região')['Valor Pago'].sum().sort_values(ascending=False)
print(total_por_regiao)

print("\nAs duas regiões que mais receberam recursos:")
print(total_por_regiao.head(2))

Região
Nordeste        4.644279e+10
Sudeste         3.426218e+10
Múltiplo        3.010563e+10
Norte           2.034243e+10
Sul             1.648147e+10
Centro-Oeste    1.037094e+10
Nacional        7.885927e+09
Exterior        2.572317e+07
Name: Valor Pago, dtype: float64

As duas regiões que mais receberam recursos:
Região
Nordeste    4.644279e+10
Sudeste     3.426218e+10
Name: Valor Pago, dtype: float64


As duas regiões que mais receberam recursos foram Nordeste e Sudeste.

## 13) Calcule a **quantidade** de emendas por **Nome Função**. Mostre as 5 funções com mais emendas.

In [33]:
contagem_funcao = df['Nome Função'].value_counts()
print(contagem_funcao.head(5))

Nome Função
SAÚDE                 37039
URBANISMO              8655
EDUCAÇÃO               8101
AGRICULTURA            4938
ENCARGOS ESPECIAIS     4610
Name: count, dtype: int64


As 5 funções com mais emendas são: Saúde(37039), Urbanismo(8655), Educação(8101), Agricultura(4938) e Encargos Especiais(4610).

## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [34]:
df['Saldo a Pagar'] = df['Valor Empenhado'] - df['Valor Pago']

print(df[['Valor Empenhado', 'Valor Pago', 'Saldo a Pagar']].head())

   Valor Empenhado  Valor Pago  Saldo a Pagar
0        700000.00         0.0      700000.00
1        487500.00         0.0      487500.00
2        100000.00         0.0      100000.00
3        243750.00         0.0      243750.00
4       1658333.94         0.0     1658333.94


A coluna Saldo a Pagar demonstra os resultados obtidos a partir da subtração entre a coluna Valor Empenhado e Valor Pago.

## 15) Elimine a(s) coluna(s) do dataframe que eestá totalmente vazia.

In [56]:
df = df.drop(columns=['Saldo a Pagar'])

In [57]:
df.columns

Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos', 'Código da Emenda_1'],
      dtype='str')

A coluna Saldo a Pagar foi eliminada utilizando o comando df.drop(columns=['Coluna_X'])

## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.

In [ ]:
df.to_json('emendas_transformadas.json', orient='records', force_ascii=False, indent=4)